OpenSim/Common/APDMDataReader.cpp

```diff
@@ -0,0 +1,243 @@
+#include <fstream>
+#include "Simbody.h"
+#include "Exception.h"
+#include "FileAdapter.h"
+#include "TimeSeriesTable.h"
+#include "APDMDataReader.h"
+
+namespace OpenSim {
+
+const std::vector<std::string> APDMDataReader::acceleration_labels{
+        "/Acceleration/X", "/Acceleration/Y", "/Acceleration/Z"
+}; 
+const std::vector<std::string> APDMDataReader::angular_velocity_labels{
+    "/Angular Velocity/X", "/Angular Velocity/Y","/Angular Velocity/Z"
+};
+const std::vector<std::string> APDMDataReader::magnetic_heading_labels{
+    "/Magnetic Field/X", "/Magnetic Field/Y","/Magnetic Field/Z"
+};
+const std::vector<std::string> APDMDataReader::orientation_labels{
+    "/Orientation/Scalar", "/Orientation/X", "/Orientation/Y","/Orientation/Z"
+};
+
+const std::string APDMDataReader::TimeLabel{ "Time" };
+
+APDMDataReader* APDMDataReader::clone() const {
+    return new APDMDataReader{*this};
+}
+
+DataAdapter::OutputTables 
+APDMDataReader::extendRead(const std::string& fileName) const {
+
+    OPENSIM_THROW_IF(fileName.empty(),
+        EmptyFileName);
+
+    std::ifstream in_stream{ fileName };
+    OPENSIM_THROW_IF(!in_stream.good(),
+        FileDoesNotExist,
+        fileName);
+
+    OPENSIM_THROW_IF(in_stream.peek() == std::ifstream::traits_type::eof(),
+        FileIsEmpty,
+        fileName);
+
+    std::vector<std::string> labels;
+    // files specified by prefix + file name exist
+    double dataRate = SimTK::NaN;
+    std::vector<int> accIndex;
+    std::vector<int>  gyroIndex;
+    std::vector<int>  magIndex;
+    std::vector<int>  orientationsIndex;
+
+    int n_imus = _settings.getProperty_ExperimentalSensors().size();
+    int last_size = 1024;
+    // Will read data into pre-allocated Matrices in-memory rather than appendRow
+    // on the fly to avoid the overhead of 
```

Comment:
This comment is incomplete .... overhead of what?

Reply:
Thanks @jenhicks I'll update the comment